hydro_lang/src/graph/debug.rs

```diff
@@ -152,3 +109,295 @@ where
+    let original_size = json_content.len();
+    
+    // Skip compression for small JSON
+    if !config.enable_compression || original_size < config.min_compression_size {
+        let encoded = encode_base64_url_safe(json_content.as_bytes());
+        return (encoded, false, 1.0);
+    }
+    
+    // Try compression
+    match compress_json(json_content) {
+        Ok(compressed) => {
+            let compressed_size = compressed.len();
+            let ratio = original_size as f64 / compressed_size as f64;
+            
+            // Only use compression if it actually reduces size
+            if compressed_size < original_size {
+                let encoded = encode_base64_url_safe(&compressed);
+                println!("📦 Compressed JSON: {} bytes → {} bytes (ratio: {:.2}x)", 
+                         original_size, compressed_size, ratio);
+                (encoded, true, ratio)
+            } else {
+                // Compression didn't help, use uncompressed
+                let encoded = encode_base64_url_safe(json_content.as_bytes());
+                (encoded, false, 1.0)
+            }
+        }
+        Err(e) => {
+            // Compression failed, fall back to uncompressed
+            println!("⚠️  Compression failed: {}, using uncompressed", e);
+            let encoded = encode_base64_url_safe(json_content.as_bytes());
+            (encoded, false, 1.0)
+        }
+    }
+}
+
+/// Calculate the total URL length for a given encoded data and parameter name.
+/// Returns the total length including base URL, parameter name, and encoded data.
+#[cfg(feature = "viz")]
+fn calculate_url_length(base_url: &str, param_name: &str, encoded_data: &str) -> usize {
+    // Format: base_url#param_name=encoded_data
+    base_url.len() + 1 + param_name.len() + 1 + encoded_data.len()
+}
+
+/// Generate a URL for the visualizer with the given JSON content.
+/// Automatically chooses between compressed and uncompressed encoding based on URL length.
+/// Returns (url, is_compressed) or None if the URL would be too long.
+#[cfg(feature = "viz")]
+fn generate_visualizer_url(
+    json_content: &str,
+    config: &VisualizerConfig,
+) -> Option<(String, bool)> {
+    let (encoded_data, is_compressed, _ratio) = try_compress_and_encode(json_content, config);
+    
+    // Determine parameter name based on compression
+    let param_name = if is_compressed { "compressed" } else { "data" };
+    
+    // Calculate total URL length
+    let url_length = calculate_url_length(&config.base_url, param_name, &encoded_data);
+    
+    println!("🔗 URL length: {} characters (max: {})", url_length, config.max_url_length);
+    
+    // Check if URL is within length limit
+    if url_length <= config.max_url_length {
+        let url = format!("{}#{}={}", config.base_url, param_name, encoded_data);
+        Some((url, is_compressed))
+    } else {
+        // If compressed URL is too long, try uncompressed as a fallback
+        if is_compressed {
```

Comment:
Wouldn't the uncompressed always be larger than compressed? So could we eliminate this block?

Reply:
Still open it seems.